File: test/cases/happy/elixir/lib/worker.ex
# Copyright 2026 Chengxi Luo
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

defmodule HappyWorker do
  @moduledoc """
  Happy-path Elixir worker: `all_types` in binary and json.

  Go is the --ref language and owns both byte layouts; see
  test/cases/happy/go/type.go. The json format must match Go's encoding/json
  byte-for-byte (with SetEscapeHTML(false)): schema field order, map keys in
  byte order (Elixir binary comparison), []byte as base64, floats in shortest
  form without a trailing .0, and U+2028/U+2029 escaped (Go escapes those
  unconditionally).
  """

  @status_variants ["pending", "paid", "shipped", "delivered", "cancelled"]

  def main(_args) do
    WorkerLib.run_suite(%{
      "all_types" => %{
        "binary" => {&binary_serialize/1, &binary_deserialize/1},
        "json" => {&json_serialize/1, &json_deserialize/1}
      }
    })
  end

  defp status_ordinal(s) do
    case Enum.find_index(@status_variants, &(&1 == s)) do
      nil -> raise "unknown status #{inspect(s)}"
      i -> i
    end
  end

  # --- binary format ---------------------------------------------------------

  defp len_str(s), do: <<byte_size(s)::little-32, s::binary>>

  defp binary_serialize(fm) do
    list = fm["list"]
    opt = case fm["optional"] do
      nil -> <<0>>
      s -> <<1, len_str(s)::binary>>
    end
    [a0, a1, a2, a3] = fm["array"]
    p = fm["struct"]
    m = fm["map"]
    ms = fm["map_struct"]

    map_bin =
      m
      |> Map.keys()
      |> Enum.sort()
      |> Enum.map(fn k -> <<len_str(k)::binary, m[k]::little-32>> end)
      |> IO.iodata_to_binary()

    ms_bin =
      ms
      |> Map.keys()
      |> Enum.sort()
      |> Enum.map(fn k ->
        t = ms[k]
        <<len_str(k)::binary, len_str(t["name"])::binary, t["weight"]::little-32>>
      end)
      |> IO.iodata_to_binary()

    <<
      fm["uint8"],
      fm["uint16"]::little-16,
      fm["uint32"]::little-32,
      fm["uint64"]::little-64,
      fm["int8"]::signed-little-8,
      fm["int16"]::signed-little-16,
      fm["int32"]::signed-little-32,
      fm["int64"]::signed-little-64,
      fm["float32"]::float-little-32,
      fm["float64"]::float-little-64,
      if(fm["bool"], do: 1, else: 0),
      len_str(fm["string"])::binary,
      byte_size(fm["bytes"])::little-32,
      fm["bytes"]::binary,
      length(list)::little-32,
      Enum.map_join(list, &len_str/1)::binary,
      opt::binary,
      a0::little-32, a1::little-32, a2::little-32, a3::little-32,
      p["x"]::signed-little-32,
      p["y"]::signed-little-32,
      p["z"]::signed-little-32,
      len_str(p["name"])::binary,
      map_size(m)::little-32,
      map_bin::binary,
      map_size(ms)::little-32,
      ms_bin::binary,
      status_ordinal(fm["status"])
    >>
  end

  defp take_str(<<n::little-32, rest::binary>>) do
    <<s::binary-size(n), rest::binary>> = rest
    {s, rest}
  end

  defp take_strs(rest, 0, acc), do: {Enum.reverse(acc), rest}

  defp take_strs(rest, n, acc) do
    {s, rest} = take_str(rest)
    take_strs(rest, n - 1, [s | acc])
  end

  defp take_map(rest, 0, acc), do: {acc, rest}

  defp take_map(rest, n, acc) do
    {k, rest} = take_str(rest)
    <<v::little-32, rest::binary>> = rest
    take_map(rest, n - 1, Map.put(acc, k, v))
  end

  defp take_map_struct(rest, 0, acc), do: {acc, rest}

  defp take_map_struct(rest, n, acc) do
    {k, rest} = take_str(rest)
    {name, rest} = take_str(rest)
    <<weight::little-32, rest::binary>> = rest
    take_map_struct(rest, n - 1, Map.put(acc, k, %{"name" => name, "weight" => weight}))
  end

  defp binary_deserialize(data) do
    <<
      u8,
      u16::little-16,
      u32::little-32,
      u64::little-64,
      i8::signed-little-8,
      i16::signed-little-16,
      i32::signed-little-32,
      i64::signed-little-64,
      f32::float-little-32,
      f64::float-little-64,
      bool_byte,
      rest::binary
    >> = data

    {string, rest} = take_str(rest)
    <<nbytes::little-32, bytes::binary-size(nbytes), rest::binary>> = rest
    <<nlist::little-32, rest::binary>> = rest
    {list, rest} = take_strs(rest, nlist, [])

    <<has_opt, rest::binary>> = rest

    {optional, rest} =
      if has_opt == 0 do
        {nil, rest}
      else
        take_str(rest)
      end

    <<a0::little-32, a1::little-32, a2::little-32, a3::little-32, rest::binary>> = rest
    <<px::signed-little-32, py::signed-little-32, pz::signed-little-32, rest::binary>> = rest
    {pname, rest} = take_str(rest)

    <<nmap::little-32, rest::binary>> = rest
    {m, rest} = take_map(rest, nmap, %{})
    <<nms::little-32, rest::binary>> = rest
    {ms, rest} = take_map_struct(rest, nms, %{})

    <<ord>> = rest

    status = Enum.at(@status_variants, ord) || raise "status ordinal #{ord} out of range"

    %{
      "uint8" => u8,
      "uint16" => u16,
      "uint32" => u32,
      "uint64" => u64,
      "int8" => i8,
      "int16" => i16,
      "int32" => i32,
      "int64" => i64,
      "float32" => f32,
      "float64" => f64,
      "bool" => bool_byte != 0,
      "string" => string,
      "bytes" => bytes,
      "list" => list,
      "optional" => optional,
      "array" => [a0, a1, a2, a3],
      "struct" => %{"x" => px, "y" => py, "z" => pz, "name" => pname},
      "map" => m,
      "map_struct" => ms,
      "status" => status
    }
  end

  # --- json format -------------------------------------------------------------

  # Go's encoding/json string escaping with SetEscapeHTML(false): only \n, \r,
  # \t are named (\b and \f become \u00xx escapes), and U+2028/U+2029 are
  # escaped unconditionally (UTF-8: E2 80 A8 / E2 80 A9).
  defp go_str(s), do: IO.iodata_to_binary([?", go_str_bytes(s, []), ?"])

  defp go_str_bytes(<<>>, acc), do: Enum.reverse(acc)

  defp go_str_bytes(<<0xE2, 0x80, c, rest::binary>>, acc) when c in [0xA8, 0xA9] do
    esc = if c == 0xA8, do: "\\u2028", else: "\\u2029"
    go_str_bytes(rest, [esc | acc])
  end

  defp go_str_bytes(<<?", rest::binary>>, acc), do: go_str_bytes(rest, ["\\\"" | acc])
  defp go_str_bytes(<<?\\, rest::binary>>, acc), do: go_str_bytes(rest, ["\\\\" | acc])
  defp go_str_bytes(<<?\n, rest::binary>>, acc), do: go_str_bytes(rest, ["\\n" | acc])
  defp go_str_bytes(<<?\r, rest::binary>>, acc), do: go_str_bytes(rest, ["\\r" | acc])
  defp go_str_bytes(<<?\t, rest::binary>>, acc), do: go_str_bytes(rest, ["\\t" | acc])

  defp go_str_bytes(<<c, rest::binary>>, acc) when c < 0x20 do
    esc = "\\u" <> String.pad_leading(Integer.to_string(c, 16), 4, "0")
    go_str_bytes(rest, [String.downcase(esc) | acc])
  end

  defp go_str_bytes(<<c, rest::binary>>, acc), do: go_str_bytes(rest, [<<c>> | acc])

  # Go prints floats in shortest round-trip form without a trailing .0.
  defp go_f64(v) do
    s = :erlang.float_to_binary(v, [:short])
    if String.ends_with?(s, ".0"), do: binary_part(s, 0, byte_size(s) - 2), else: s
  end

  # Shortest decimal that round-trips through float32 (v is the f64 widening).
  defp go_f32(v) do
    bits = <<v::float-little-32>>
    Enum.find_value(0..9, fn d ->
      s = :erlang.float_to_binary(v, decimals: d)
      {f, ""} = Float.parse(s)

      if <<f::float-little-32>> == bits do
        s
        |> String.replace(~r/\.?0+$/, "")
        |> case do
          "" -> "0"
          "-" -> "-0"
          out -> out
        end
      end
    end) || go_f64(v)
  end

  defp json_serialize(fm) do
    m = fm["map"]
    ms = fm["map_struct"]
    p = fm["struct"]

    map_json =
      m
      |> Map.keys()
      |> Enum.sort()
      |> Enum.map_join(",", fn k -> "#{go_str(k)}:#{m[k]}" end)

    ms_json =
      ms
      |> Map.keys()
      |> Enum.sort()
      |> Enum.map_join(",", fn k ->
        t = ms[k]
        "#{go_str(k)}:{\"name\":#{go_str(t["name"])},\"weight\":#{t["weight"]}}"
      end)

    IO.iodata_to_binary([
      "{\"uint8\":#{fm["uint8"]}",
      ",\"uint16\":#{fm["uint16"]}",
      ",\"uint32\":#{fm["uint32"]}",
      ",\"uint64\":#{fm["uint64"]}",
      ",\"int8\":#{fm["int8"]}",
      ",\"int16\":#{fm["int16"]}",
      ",\"int32\":#{fm["int32"]}",
      ",\"int64\":#{fm["int64"]}",
      ",\"float32\":#{go_f32(fm["float32"])}",
      ",\"float64\":#{go_f64(fm["float64"])}",
      ",\"bool\":#{fm["bool"]}",
      ",\"string\":#{go_str(fm["string"])}",
      ",\"bytes\":\"#{Base.encode64(fm["bytes"])}\"",
      ",\"list\":[#{Enum.map_join(fm["list"], ",", &go_str/1)}]",
      ",\"optional\":#{if fm["optional"] == nil, do: "null", else: go_str(fm["optional"])}",
      ",\"array\":[#{Enum.join(fm["array"], ",")}]",
      ",\"struct\":{\"x\":#{p["x"]},\"y\":#{p["y"]},\"z\":#{p["z"]},\"name\":#{go_str(p["name"])}}",
      ",\"map\":{#{map_json}}",
      ",\"map_struct\":{#{ms_json}}",
      ",\"status\":#{go_str(fm["status"])}}"
    ])
  end

  # Jason decodes integer literals as integers; float fields must come back as
  # floats (Go emits e.g. `0` for a zero float64).
  defp to_float(v) when is_integer(v), do: v * 1.0
  defp to_float(v), do: v

  defp json_deserialize(data) do
    v = Jason.decode!(data)

    %{
      "uint8" => v["uint8"],
      "uint16" => v["uint16"],
      "uint32" => v["uint32"],
      "uint64" => v["uint64"],
      "int8" => v["int8"],
      "int16" => v["int16"],
      "int32" => v["int32"],
      "int64" => v["int64"],
      "float32" => to_float(v["float32"]),
      "float64" => to_float(v["float64"]),
      "bool" => v["bool"],
      "string" => v["string"],
      "bytes" => Base.decode64!(v["bytes"]),
      "list" => v["list"],
      "optional" => v["optional"],
      "array" => v["array"],
      "struct" => v["struct"],
      "map" => v["map"],
      "map_struct" => v["map_struct"],
      "status" => v["status"]
    }
  end
end

File: test/cases/happy/java/src/main/java/Worker.java
/*
 * Copyright 2026 Chengxi Luo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Happy-path Java worker: {@code all_types} in binary and json.
 *
 * <p>Go is the --ref language and owns both byte layouts; see
 * test/cases/happy/go/type.go. The json format must match Go's encoding/json
 * byte-for-byte (with SetEscapeHTML(false)): schema field order, map keys in
 * UTF-8 byte order, []byte as base64, floats in shortest form without a
 * trailing .0, and U+2028/U+2029 escaped (Go escapes those unconditionally).
 *
 * <p>Java has no unsigned types: u32 lives in an int and u64 in a long as bit
 * patterns, printed with Integer/Long.toUnsignedString.
 */

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.serify.WorkerLib;
import io.serify.WorkerLib.FieldMap;
import io.serify.WorkerLib.FormatPair;
import io.serify.WorkerLib.TypeEntry;

import java.io.ByteArrayOutputStream;
import java.math.BigDecimal;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class Worker {

    private static final List<String> STATUS_VARIANTS =
            List.of("pending", "paid", "shipped", "delivered", "cancelled");

    private static byte statusOrdinal(String s) {
        int i = STATUS_VARIANTS.indexOf(s);
        if (i < 0) throw new IllegalArgumentException("unknown status \"" + s + "\"");
        return (byte) i;
    }

    /** UTF-8 byte order (== code-point order), not UTF-16 code-unit order. */
    private static int byteCompare(String a, String b) {
        byte[] ab = a.getBytes(StandardCharsets.UTF_8);
        byte[] bb = b.getBytes(StandardCharsets.UTF_8);
        int n = Math.min(ab.length, bb.length);
        for (int i = 0; i < n; i++) {
            int d = (ab[i] & 0xff) - (bb[i] & 0xff);
            if (d != 0) return d;
        }
        return ab.length - bb.length;
    }

    // --- binary format -------------------------------------------------------

    private static void writeLenStr(ByteArrayOutputStream out, String s) {
        byte[] b = s.getBytes(StandardCharsets.UTF_8);
        out.writeBytes(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(b.length).array());
        out.writeBytes(b);
    }

    private static String readLenStr(ByteBuffer buf) {
        byte[] b = new byte[buf.getInt()];
        buf.get(b);
        return new String(b, StandardCharsets.UTF_8);
    }

    private static byte[] binarySerialize(FieldMap fm) {
        ByteArrayOutputStream out = new ByteArrayOutputStream();

        ByteBuffer head = ByteBuffer.allocate(43).order(ByteOrder.LITTLE_ENDIAN);
        head.put(fm.getU8("uint8"));
        head.putShort(fm.getU16("uint16"));
        head.putInt(fm.getU32("uint32"));
        head.putLong(fm.getU64("uint64"));
        head.put(fm.getI8("int8"));
        head.putShort(fm.getI16("int16"));
        head.putInt(fm.getI32("int32"));
        head.putLong(fm.getI64("int64"));
        head.putFloat(fm.getF32("float32"));
        head.putDouble(fm.getF64("float64"));
        head.put((byte) (fm.getBool("bool") ? 1 : 0));
        out.writeBytes(head.array());

        writeLenStr(out, fm.getString("string"));

        byte[] raw = fm.getBytes("bytes");
        out.writeBytes(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(raw.length).array());
        out.writeBytes(raw);

        List<String> list = fm.getListString("list");
        out.writeBytes(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(list.size()).array());
        for (String s : list) writeLenStr(out, s);

        var opt = fm.getOptionalString("optional");
        if (opt.isEmpty()) {
            out.write(0);
        } else {
            out.write(1);
            writeLenStr(out, opt.get());
        }

        ByteBuffer arr = ByteBuffer.allocate(16).order(ByteOrder.LITTLE_ENDIAN);
        for (int n : fm.getListU32("array")) arr.putInt(n);
        out.writeBytes(arr.array());

        FieldMap p = fm.getStruct("struct");
        ByteBuffer pt = ByteBuffer.allocate(12).order(ByteOrder.LITTLE_ENDIAN);
        pt.putInt(p.getI32("x"));
        pt.putInt(p.getI32("y"));
        pt.putInt(p.getI32("z"));
        out.writeBytes(pt.array());
        writeLenStr(out, p.getString("name"));

        Map<String, Object> m = fm.getMap("map");
        List<String> keys = new ArrayList<>(m.keySet());
        keys.sort(Worker::byteCompare);
        out.writeBytes(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(keys.size()).array());
        for (String k : keys) {
            writeLenStr(out, k);
            int v = ((Number) m.get(k)).intValue();
            out.writeBytes(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(v).array());
        }

        Map<String, Object> ms = fm.getMap("map_struct");
        List<String> mkeys = new ArrayList<>(ms.keySet());
        mkeys.sort(Worker::byteCompare);
        out.writeBytes(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(mkeys.size()).array());
        for (String k : mkeys) {
            FieldMap t = (FieldMap) ms.get(k);
            writeLenStr(out, k);
            writeLenStr(out, t.getString("name"));
            out.writeBytes(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(t.getU32("weight")).array());
        }

        out.write(statusOrdinal(fm.getString("status")));
        return out.toByteArray();
    }

    private static FieldMap binaryDeserialize(byte[] data) {
        ByteBuffer buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
        FieldMap fm = new FieldMap();
        fm.setU8("uint8", buf.get());
        fm.setU16("uint16", buf.getShort());
        fm.setU32("uint32", buf.getInt());
        fm.setU64("uint64", buf.getLong());
        fm.setI8("int8", buf.get());
        fm.setI16("int16", buf.getShort());
        fm.setI32("int32", buf.getInt());
        fm.setI64("int64", buf.getLong());
        fm.setF32("float32", buf.getFloat());
        fm.setF64("float64", buf.getDouble());
        fm.setBool("bool", buf.get() != 0);
        fm.setString("string", readLenStr(buf));

        byte[] raw = new byte[buf.getInt()];
        buf.get(raw);
        fm.setBytes("bytes", raw);

        int nlist = buf.getInt();
        List<String> list = new ArrayList<>(nlist);
        for (int i = 0; i < nlist; i++) list.add(readLenStr(buf));
        fm.setListString("list", list);

        fm.setOptionalString("optional", buf.get() != 0 ? readLenStr(buf) : null);

        var arr = new ArrayList<Integer>();
        for (int i = 0; i < 4; i++) arr.add(buf.getInt());
        fm.setListU32("array", arr);

        FieldMap p = new FieldMap();
        p.setI32("x", buf.getInt());
        p.setI32("y", buf.getInt());
        p.setI32("z", buf.getInt());
        p.setString("name", readLenStr(buf));
        fm.setStruct("struct", p);

        Map<String, Object> m = new HashMap<>();
        int nmap = buf.getInt();
        for (int i = 0; i < nmap; i++) {
            String k = readLenStr(buf);
            m.put(k, buf.getInt());
        }
        fm.setMap("map", m);

        Map<String, Object> ms = new HashMap<>();
        int nms = buf.getInt();
        for (int i = 0; i < nms; i++) {
            String k = readLenStr(buf);
            FieldMap t = new FieldMap();
            t.setString("name", readLenStr(buf));
            t.setU32("weight", buf.getInt());
            ms.put(k, t);
        }
        fm.setMap("map_struct", ms);

        int ord = buf.get() & 0xff;
        if (ord >= STATUS_VARIANTS.size()) {
            throw new IllegalArgumentException("status ordinal " + ord + " out of range");
        }
        fm.setString("status", STATUS_VARIANTS.get(ord));
        return fm;
    }

    // --- json format -----------------------------------------------------------

    /**
     * Go's encoding/json string escaping with SetEscapeHTML(false): only \n,
     * \r, \t are named (\b and \f become u00xx escapes), and U+2028/U+2029 are
     * escaped unconditionally.
     */
    private static String goStr(String s) {
        StringBuilder sb = new StringBuilder("\"");
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            switch (ch) {
                case '"' -> sb.append("\\\"");
                case '\\' -> sb.append("\\\\");
                case '\n' -> sb.append("\\n");
                case '\r' -> sb.append("\\r");
                case '\t' -> sb.append("\\t");
                case '\u2028' -> sb.append("\\u2028");
                case '\u2029' -> sb.append("\\u2029");
                default -> {
                    if (ch < 0x20) {
                        sb.append(String.format("\\u%04x", (int) ch));
                    } else {
                        sb.append(ch);
                    }
                }
            }
        }
        return sb.append('"').toString();
    }

    /**
     * Go prints floats in shortest round-trip form without a trailing .0 and
     * (in this value range) without an exponent. Double.toString gives the
     * digits; BigDecimal removes Java's E-notation for magnitudes >= 1e7.
     */
    private static String goF64(double v) {
        String plain = new BigDecimal(Double.toString(v)).toPlainString();
        return plain.endsWith(".0") ? plain.substring(0, plain.length() - 2) : plain;
    }

    private static String goF32(float v) {
        String plain = new BigDecimal(Float.toString(v)).toPlainString();
        return plain.endsWith(".0") ? plain.substring(0, plain.length() - 2) : plain;
    }

    private static byte[] jsonSerialize(FieldMap fm) {
        StringBuilder sb = new StringBuilder("{");
        sb.append("\"uint8\":").append(Byte.toUnsignedInt(fm.getU8("uint8")));
        sb.append(",\"uint16\":").append(Short.toUnsignedInt(fm.getU16("uint16")));
        sb.append(",\"uint32\":").append(Integer.toUnsignedString(fm.getU32("uint32")));
        sb.append(",\"uint64\":").append(Long.toUnsignedString(fm.getU64("uint64")));
        sb.append(",\"int8\":").append(fm.getI8("int8"));
        sb.append(",\"int16\":").append(fm.getI16("int16"));
        sb.append(",\"int32\":").append(fm.getI32("int32"));
        sb.append(",\"int64\":").append(fm.getI64("int64"));
        sb.append(",\"float32\":").append(goF32(fm.getF32("float32")));
        sb.append(",\"float64\":").append(goF64(fm.getF64("float64")));
        sb.append(",\"bool\":").append(fm.getBool("bool"));
        sb.append(",\"string\":").append(goStr(fm.getString("string")));
        sb.append(",\"bytes\":\"").append(Base64.getEncoder().encodeToString(fm.getBytes("bytes"))).append('"');

        sb.append(",\"list\":[");
        List<String> list = fm.getListString("list");
        for (int i = 0; i < list.size(); i++) {
            if (i > 0) sb.append(',');
            sb.append(goStr(list.get(i)));
        }
        sb.append(']');

        var opt = fm.getOptionalString("optional");
        sb.append(",\"optional\":").append(opt.isEmpty() ? "null" : goStr(opt.get()));

        sb.append(",\"array\":[");
        var arr = fm.getListU32("array");
        for (int i = 0; i < arr.size(); i++) {
            if (i > 0) sb.append(',');
            sb.append(Integer.toUnsignedString(arr.get(i)));
        }
        sb.append(']');

        FieldMap p = fm.getStruct("struct");
        sb.append(",\"struct\":{\"x\":").append(p.getI32("x"))
          .append(",\"y\":").append(p.getI32("y"))
          .append(",\"z\":").append(p.getI32("z"))
          .append(",\"name\":").append(goStr(p.getString("name"))).append('}');

        Map<String, Object> m = fm.getMap("map");
        List<String> keys = new ArrayList<>(m.keySet());
        keys.sort(Worker::byteCompare);
        sb.append(",\"map\":{");
        for (int i = 0; i < keys.size(); i++) {
            if (i > 0) sb.append(',');
            String k = keys.get(i);
            sb.append(goStr(k)).append(':')
              .append(Integer.toUnsignedString(((Number) m.get(k)).intValue()));
        }
        sb.append('}');

        Map<String, Object> ms = fm.getMap("map_struct");
        List<String> mkeys = new ArrayList<>(ms.keySet());
        mkeys.sort(Worker::byteCompare);
        sb.append(",\"map_struct\":{");
        for (int i = 0; i < mkeys.size(); i++) {
            if (i > 0) sb.append(',');
            String k = mkeys.get(i);
            FieldMap t = (FieldMap) ms.get(k);
            sb.append(goStr(k)).append(":{\"name\":").append(goStr(t.getString("name")))
              .append(",\"weight\":").append(Integer.toUnsignedString(t.getU32("weight"))).append('}');
        }
        sb.append('}');

        sb.append(",\"status\":").append(goStr(fm.getString("status")));
        sb.append('}');
        return sb.toString().getBytes(StandardCharsets.UTF_8);
    }

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static FieldMap jsonDeserialize(byte[] data) throws Exception {
        JsonNode v = MAPPER.readTree(data);

        FieldMap fm = new FieldMap();
        fm.setU8("uint8", (byte) v.get("uint8").asInt());
        fm.setU16("uint16", (short) v.get("uint16").asInt());
        fm.setU32("uint32", (int) v.get("uint32").asLong());
        fm.setU64("uint64", v.get("uint64").bigIntegerValue().longValue());
        fm.setI8("int8", (byte) v.get("int8").asInt());
        fm.setI16("int16", (short) v.get("int16").asInt());
        fm.setI32("int32", v.get("int32").asInt());
        fm.setI64("int64", v.get("int64").asLong());
        fm.setF32("float32", (float) v.get("float32").asDouble());
        fm.setF64("float64", v.get("float64").asDouble());
        fm.setBool("bool", v.get("bool").asBoolean());
        fm.setString("string", v.get("string").asText());
        fm.setBytes("bytes", Base64.getDecoder().decode(v.get("bytes").asText()));

        List<String> list = new ArrayList<>();
        v.get("list").forEach(x -> list.add(x.asText()));
        fm.setListString("list", list);

        JsonNode opt = v.get("optional");
        fm.setOptionalString("optional", opt.isNull() ? null : opt.asText());

        var arr = new ArrayList<Integer>();
        JsonNode ja = v.get("array");
        for (int i = 0; i < 4; i++) arr.add((int) ja.get(i).asLong());
        fm.setListU32("array", arr);

        JsonNode st = v.get("struct");
        FieldMap p = new FieldMap();
        p.setI32("x", st.get("x").asInt());
        p.setI32("y", st.get("y").asInt());
        p.setI32("z", st.get("z").asInt());
        p.setString("name", st.get("name").asText());
        fm.setStruct("struct", p);

        Map<String, Object> m = new HashMap<>();
        v.get("map").fields().forEachRemaining(e -> m.put(e.getKey(), (int) e.getValue().asLong()));
        fm.setMap("map", m);

        Map<String, Object> ms = new HashMap<>();
        v.get("map_struct").fields().forEachRemaining(e -> {
            FieldMap t = new FieldMap();
            t.setString("name", e.getValue().get("name").asText());
            t.setU32("weight", (int) e.getValue().get("weight").asLong());
            ms.put(e.getKey(), t);
        });
        fm.setMap("map_struct", ms);

        fm.setString("status", v.get("status").asText());
        return fm;
    }

    public static void main(String[] args) {
        WorkerLib.runSuite(Map.of(
                "all_types", TypeEntry.formats(Map.of(
                        "binary", new FormatPair(Worker::binarySerialize, Worker::binaryDeserialize),
                        "json", new FormatPair(fm -> {
                            try {
                                return jsonSerialize(fm);
                            } catch (Exception e) {
                                throw new RuntimeException(e);
                            }
                        }, data -> {
                            try {
                                return jsonDeserialize(data);
                            } catch (Exception e) {
                                throw new RuntimeException(e);
                            }
                        })))));
    }

    private Worker() {}
}
